src/connectors/impls/clickhouse.rs

```diff
@@ -0,0 +1,308 @@
+// that most of the logic is already written once it is available.
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+enum DummySqlType {
+    Array(Box<DummySqlType>),
+    Nullable(Box<DummySqlType>),
+
+    Int64,
+    UInt64,
+    String,
+}
+
+impl DummySqlType {
+    fn as_non_nullable(&self) -> &DummySqlType {
+        match self {
+            DummySqlType::Nullable(inner_type) => inner_type.as_ref(),
+            non_nullable => non_nullable,
+        }
+    }
+
+    fn wrap_if_nullable(&self, value: clickhouse_rs::types::Value) -> clickhouse_rs::types::Value {
+        match self {
+            DummySqlType::Nullable(_) => panic!("Nullable values can not be constructed :("),
+            _ => value,
+        }
+    }
+}
+
+fn clickhouse_value_of(
+    cell: &Value,
+    expected_type: &DummySqlType,
+) -> Result<clickhouse_rs::types::Value> {
+    use clickhouse_rs::types;
+
+    match (cell, expected_type) {
+        (Value::Static(value), _) => {
+            if matches!(
+                (value, expected_type),
+                (StaticNode::Null, DummySqlType::Nullable(_))
+            ) {
+                // Null can be of any type, as long as it is allowed by the
+                // schema.
+
+                // The situation is quite sad. The problem is that
+                // clickhouse_rs::types::Value::Nullable wraps an Either [1],
+                // which defined in clickhouse itself and is not publicly
+                // available [2]. Once [3] is merged, Value should be fully
+                // constructible again!
+                //
+                // [1]: https://docs.rs/clickhouse-rs/1.0.0-alpha.1/clickhouse_rs/types/enum.Value.html#variant.Nullable
+                // [2]: https://docs.rs/clickhouse-rs/0.1.21/src/clickhouse_rs/types/either.rs.html#4
+                // [3]: https://github.com/suharev7/clickhouse-rs/pull/171
+                panic!("Null value can't be constructed for now :(");
+            }
+
+            let value_as_non_null = match (value, expected_type.as_non_nullable()) {
+                // These are the *obvious* translations. No cast is required,
+                // Not much to say here.
+                (StaticNode::U64(v), DummySqlType::UInt64) => types::Value::UInt64(*v),
+                (StaticNode::I64(v), DummySqlType::Int64) => types::Value::Int64(*v),
+
+                // Booleans can be converted to integers (true = 1, false = 0).
+                (StaticNode::Bool(b), DummySqlType::UInt64) => types::Value::UInt64(u64::from(*b)),
+                (StaticNode::Bool(b), DummySqlType::Int64) => types::Value::Int64(i64::from(*b)),
+
+                _ => todo!(),
+            };
+
+            Ok(expected_type.wrap_if_nullable(value_as_non_null))
+        }
+
+        (Value::Array(values), DummySqlType::Array(expected_inner_type)) => values
+            .iter()
+            .map(|value| clickhouse_value_of(value, expected_inner_type))
```

Comment:
We don't check that all the elements of the array have the same type. Instead, we check that each element of the array is convertible to the expected array type.
I fail to see a situation where this could be a problem, but I am not sure it is the best way to proceed.

Reply:
That's probably not only fine but the better approach. For example we don't give users a easy way to differentiate between a u8 and a u64 so using "is the same type" instead of "converts to the same type" could end up causing surprising troubles :)

Reply:
That makes complete sense, thanks!